pyteal/ast/replace.py

```diff
@@ -0,0 +1,86 @@
+from typing import cast, TYPE_CHECKING
+
+from pyteal.types import TealType, require_type
+from pyteal.errors import verifyTealVersion
+from pyteal.ir import TealOp, Op, TealBlock
+from pyteal.ast.expr import Expr
+from pyteal.ast.int import Int
+from pyteal.ast.ternaryexpr import TernaryExpr
+
+if TYPE_CHECKING:
+    from pyteal.compiler import CompileOptions
+
+
+class ReplaceExpr(Expr):
+    """An expression for replacing a section of a byte string at a given start index"""
+
+    def __init__(self, original: Expr, start: Expr, replacement: Expr) -> None:
+        super().__init__()
+
+        require_type(original, TealType.bytes)
+        require_type(start, TealType.uint64)
+        require_type(replacement, TealType.bytes)
+
+        self.original = original
+        self.start = start
+        self.replacement = replacement
+
+    # helper method for correctly populating op
+    def __get_op(self, options: "CompileOptions"):
+        s = cast(Int, self.start).value
+        if s < 2**8:
+            return Op.replace2
+        else:
+            return Op.replace3
+
+    def __teal__(self, options: "CompileOptions"):
+        if not isinstance(self.start, Int):
+            return TernaryExpr(
+                Op.replace3,
+                (TealType.bytes, TealType.uint64, TealType.bytes),
+                TealType.bytes,
+                self.original,
+                self.start,
+                self.replacement,
+            ).__teal__(options)
+
+        op = self.__get_op(options)
+
+        verifyTealVersion(
```

Comment:
This check isn't hit if `self.start` isn't an `Int`

Reply:
@jasonpaulos I had the same thought, however `TernaryExpr` checks the version during `__teal__` invocation.  If you agree, can we consider the topic resolved?

Reply:
Good point, I will resolve this then.